### PR TITLE
Ticket migrate

### PR DESCRIPTION
**Pull request by danuker**
This is the ticket to migrate tickets.

Fixes https://github.com/chevah/server/issues/1271 and https://github.com/chevah/server/issues/1267 .

GitHub REST Issue creation docs: https://docs.github.com/en/rest/reference/issues#create-an-issue

## What is not ideal:

- [x] ~~Tables or links to wiki do not work (not sure if we have any tables in a ticket).~~
- [x] Trac DB columns not used: 
  - [x] version
  - [x] milestone
- [x] Lots of Trac DB tables not used (should keep the wiki DB backup around).
  - [x] but a quick look by me did not show something important missing; more like Trac config values and history (and the wiki is already migrated).
- [x] ~~Attachments are not handled. We have just 7. Are they need to for Twisted?~~ Not in this PR. Created #14 
- [x] ~~Have not tried the interaction between the automated comments and the GitHub bot changing ticket statuses/labels. Should I censor command keywords from the comments?~~ Mentioned in #16 
- [x] ~~Add "incremental migration" to remember and skip tickets already posted. This allows safe re-runs, which will not double-post tickets. This should help us do a migration PR -> sftpplus.com tickets first.~~ Mentioned in #16 
- [x] TODO (just noticed): if repo used is fallback (server), apply component name as a tag
- [x] Assign tickets to a separate project for each Trac milestone.
  - [x] Close all created projects by default. If projects need to be open, they should be opened manually.
- [x] Convert links from the Trac ID to the GitHub ID.
- [x] Include `tickets_created.tsv` in `requestNextNumber`.
- [ ] Add a comment created later than the issue's PR, that mentions the PR, to see if it links it to the issue.
- [x] for chevah/server, instead of T2618 use trac-2618, to have links; we might keep Trac around read-only for a while.

## Formatting comparison

Trac URL | GitHub URL | tested part
-- | -- | --
https://trac.chevah.com/ticket/98 | https://github.com/chevah/trac-migration-staging/issues/14 | preformatted text
https://trac.chevah.com/ticket/637 | https://github.com/chevah/trac-migration-staging/issues/15 | embedded `{{{#!rst`
https://trac.chevah.com/ticket/822 | https://github.com/chevah/trac-migration-staging/issues/16 | horizontal lines
https://trac.chevah.com/ticket/3928 | https://github.com/chevah/trac-migration-staging/issues/25 | comments
https://trac.chevah.com/ticket/5519 | https://github.com/chevah/trac-migration-staging/issues/32 | closed status and label; **attachment is not added**; PR link is added; project is added

How to try and test the changes
===============================

reviewers: @adiroiban 

Test with `DRY_RUN=True`: generate and check proposed trac -> GH URL mappings on the real repos.
I could not do this because I don't have access to the `salt` repo, nor the current Trac dump.
If you want, test with one or more custom Trac ticket ID conditions (preferably ones with matching GitHub ID) in the `select_tickets` method, and `DRY_RUN=False`.

### CODE REVIEW

**Comment by adiroiban** [comment]
Another nice to have feature is trying to preserve the ID of trac tickets, at least for the main chevah/server repo

For example, in chevah/server we now have about 1860 PRs already created... so IDs up to 1860 are in conflict.
But at least for Trac tickets from ID 1861 and above, we can migrate them to GitHub and keep the same ID.

This will automatically enable a lot of links in the closed PR https://github.com/chevah/server/pulls?q=is%3Apr+is%3Aclosed

**Comment by danuker** [comment]
@adiroiban This is ready for review. Please look at the **What is not ideal** part especially, I am uncertain whether to handle attachments.

needs-review

File: ticket_migrate.py [hunk 1]
```diff
@@ -0,0 +1,829 @@
...
+
+import requests
+
+try:
+    import config
+except ModuleNotFoundError:
+    # In the tests, we monkeypatch this module.
+    config = None
+
+from trac2down import convert
```
Comment (danuker) [inline ticket_migrate.py:19]: Trac2down is thankfully licensed as LGPL, so linking from non-GPL code is allowed.

**Comment by adiroiban** [comment]
Thanks for the update. There are a lot of changes ... so this will take a while.

>  Trac DB columns not used: 

version is ok.
Milestone is ok as long as it is assigned to a separate project for each milestone.

Note that there are  Organization wide (multi-repo) projects 

https://github.com/orgs/chevah/projects/6

And there are repo projects 

https://github.com/chevah/server/projects/7

I see in your example that you use an repo project ... this is not what we want :)

https://github.com/chevah/trac-migration-staging/projects/1

---------------

It looks like the new features are only available for the organization-wise projects 

We want organization-wise project as they are suited for private companies and projects.

But for twisted the migration should do the migration to "milestones" as there are no projects in the Twisted open source organization.

So for this PR is ok to migrate to org-wise projects.

In a separate PR we can look at adding a config to migrate to a repo milestone.

----------

GitHub Projects are the way to go for Jira / Trac  replacement.

Milestone are a legacy GitHub thing that can still be used for very simple management.
Milestones make a lot of sense for open source project in which there is no deadline and many changes are done
directly via a PR that is accepted.
An open source user found a bug, she/he just send a PR and the PR is accepted.
No need for a ticket creation, ticket triage and prioritization, ticket sprint scheduling ... just get it done :)

----------

>  Lots of Trac DB tables not used (should keep the wiki DB backup around). 

That is ok.
We also had in the past various Trac plugins and customization that might have created their own tickets...
there was a time with full SCRUM and Kanban implemented in Trac.

--------------------

>  Attachments are not handled. We have just 7. Are they need to for Twisted?

They are not needed for us...but it would be nice to have them for Twisted .
But in a separate PR. as this PR is big already.

--------------

> Have not tried the interaction between the automated comments and the GitHub bot changing ticket statuses/labels. Should I censor command keywords from the comments?

I think that during the migration we can temporary disable the chevah robi-net automation.

We can test with PR tickets and we will see if this is required... I expect that it will be required to pause our custom GitHub automation :)

------------

What I would like to see is "an incremental" migration.
For example, when `FALLBACK_REPOSITORY = None` if there is no mapping for the component, the ticket is not migrated.
Same for `REPOSITORY_MAPPING['infrastructure'] = None` the tickets are not migrated.

This should help us do a migration PR -> sftpplus.com tickets first.

**Comment by adiroiban** [comment]
>  Tables or links to wiki do not work (not sure if we have any tables in a ticket).

We should not care about tables.

For Trac Links... we can look at this in a separate PR. 
This might be needed for Twisted repo.

For us, is ok.

If we really need trac links, we can look later and run a job to edit the comments and update the links to our GitHub wiki

File: tickets_created.tsv [hunk 2]
```diff
@@ -0,0 +1,20 @@
+Trac URL	GitHub URL	Comment
```
Comment (adiroiban) [inline tickets_created.tsv:1]: why is this added to the repo?

Is ok to have this file generated, but I think that it should stay outside of the repo.

**Review by adiroiban — CHANGES_REQUESTED** [review]
It looks good. Thanks.

Please look at adding support for a marker to skip the migration.

For us, is much better if we don't migrate the tickets is one go.

```
FALLBACK_REPOSITORY = None
or
FALLBACK_REPOSITORY = SkipMigration()
```


We should merge this as it is and do a migration for sftpplus.com and chevah/compat

Later, as we found issues, we can create separate PR and improve this.

Thanks!

File: README.rst [hunk 3]
```diff
@@ -58,3 +60,12 @@ Things that are not yet auto-converted:
 * Sub-pages listing macro `[[TitleIndex(Development/)]]`
 * Local table of content `[[PageOutline]]`
 * Manually create _Sidebar.rst and _Footer.rst GitHub wiki meta-pages.
+
+# Ticket migration
+
+1. Copy `config.py.sample` over `config.py`, and edit all the settings.
```
Comment (adiroiban) [inline README.rst:66]: I think that this step is also needed for Wiki migration

File: ticket_migrate.py [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+
+    return convert_issue_content(description)
+
+
+def convert_issue_content(text):
+    """
+    Convert TracWiki text to GitHub Markdown.
+    Ignore included images.
+    """
+    return convert(text, '')
```
Comment (adiroiban) [inline ticket_migrate.py:790]: I am not sure why we have empty string here.

When possible, is best to call the methods with named arguments

```suggestion
    return convert(text, multilines=False)
```

File: ticket_migrate.py [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+    if status in ['in_work', 'needs_changes', 'needs_merge', 'needs_review']:
+        return {status}
+
+    return set()
+
+
+def parse_body(description):
+    """
+    Parses text with squiggly-bracketed or backtick-surrounded monospace.
+    Converts the squiggly brackets to backtick brackets.
```
Comment (adiroiban) [inline ticket_migrate.py:752]:
```suggestion
    Converts the curly brackets to backtick brackets.
```

First time I hear about squiggly :)

File: ticket_migrate.py [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+            print(f"Would post to {url}:\n{comment_data}")
+            return
+
+        if not self.github_number:
+            print(f"Trac ticket {self.t_id} did not receive a GitHub number.")
+            import pdb
+            pdb.set_trace()
+
+        # Avoid crossing the rate limit (5000/hr).
+        sleep(1)
```
Comment (adiroiban) [inline ticket_migrate.py:544]: nice try :) 

Note that in the response of each header you should receive the remaining limit

File: ticket_migrate.py [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+
+        body = (
+            f"Comment by {author} at {showtime(trac_data['c_time'])}.\n"
+            f"\n"
+            f"{parse_body(trac_data['newvalue'])}"
+            )
+
+        return {'body': body}
+
+    def _submitComment(self, comment_data):
```
Comment (adiroiban) [inline ticket_migrate.py:527]: just a question... a lot of this code should not be reused in other places.

why bother with private method names ?

and `def submit_comment` or `def subbmitComment` ? )

File: ticket_migrate.py [hunk 8]
```diff
@@ -0,0 +1,829 @@
...
+            }
+
+
+def read_trac_comments():
+    """
+    Read the Trac comment data from the database.
+
+    The last version is in the `newvalue` of the `comment` field.
+
+    Example changed comment: https://trac.chevah.com/ticket/3928#comment:1
```
Comment (adiroiban) [inline ticket_migrate.py:147]: this is open source code ...but trac.chevah.com is private ...so these links are not very helpful 

we can leave them... no problem....

but we also plan to decommission trac.chevah.com so the links will be broken

and I don't know if we will setup redirection...but we might :)

File: ticket_migrate.py (hunk 4 above)
Comment (danuker) [reply ticket_migrate.py:790]: > best to call the methods with named arguments

This is `base_path`; point taken. I use an empty string because we don't plan to support image and wiki migration in this PR, yet something has to be provided as an argument.

File: ticket_migrate.py (hunk 5 above)
Comment (danuker) [reply ticket_migrate.py:752]: :D I did not realize my naming was nonstandard

File: config.py.sample [hunk 9]
```diff
@@ -10,3 +11,22 @@ TRAC_TICKET_PREFIX = 'https://trac.chevah.com/ticket/'
...
+
+# Trac ticket Component to GitHub repository mapping.
+REPOSITORY_MAPPING = {
+    'client': 'client',
+    'commons': 'commons',
+    'trac-migration-staging': 'trac-migration-staging',
+    }
+
+# GitHub repository for Trac tickets with Component not in the mapping.
+FALLBACK_REPOSITORY = 'server'
```
Comment (danuker) [inline config.py.sample:23]: > ```
> FALLBACK_REPOSITORY = None
> or
> FALLBACK_REPOSITORY = SkipMigration()
> ```

We can exclude these repositories from the `select_tickets` method. I will add a feature to skip already-submitted tickets seen in `tickets_created.tsv`.

File: ticket_migrate.py (hunk 7 above)
Comment (danuker) [reply ticket_migrate.py:527]: I don't like methods too big. It is easier to think about them if they do less.

In `submitMyComments`, the public high-level method, it iterates through tickets and calls the low-level method.
In `_submitComment`, the private low-level method, it submits a POST request if appropriate, and records that it happened.

> why bother with private method names ?

Would you just join the code into one big method? Or just make everything public?

**Comment by danuker** [comment]
I believe this is nearing completion. See #16.

Any partial run (with conditions like component being `pr`) will save the tickets and skip them on later runs.

Org-wide projects are created and closed (and also saved to avoid creating duplicates).

There is no sleep until the rate-limit is (nearly) hit, so most requests will happen as fast as possible in one thread.

Would appreciate another review.

needs-review
